Add new detection algorthim

amrsoll opened the pull request:
- Single frame detection
- Auto adjust shutter speed to detect reference markers instead of the center of the frame

New python/pip packages needed:
- ~scipy (takes forever to install)~

amrsoll committed:
use student proj marker detection

amrsoll committed:
wip - adding new detection + extra features

amrsoll committed:
Auto shutter speed compensation and verbosity

amrsoll committed:
python2 compatibility

amrsoll committed:
Merge remote-tracking branch 'origin/develop' into f_new_detection

amrsoll committed:
Stable - commented out dummy camera and gauss blur + int offset slice

amrsoll committed:
Merge remote-tracking branch 'origin/develop' into f_new_detection

amrsoll committed:
Import Marker detect algo into Plugin.

- Copy the new detection algo in camera.undistort

- Add documentation strings

- deobfuscation of the code

- Give more graceful exits

- Picture loaded from RAM instead of file

- Better at predicting when a picture is ready to analyse

- Too fast for frontend?

- Logging settings depend on debug variable

- Actually do a multicore analysis of the markers

amrsoll commented:
Still need to do before being semi-stable :

- Handle errors better (especially when start/stopping, opening lid etc...)
- get signal from front-end to know if it's ready to receive a refreshed picture
- Analytics : compare the results of the 2 algorithms
- Scale distortion matrix to accommodate higher resolution pictures
- Detect movement in order not to serve the same image twice

amrsoll committed:
Bugfix/lognames, bad ROI offset and hue tolerance.

amrsoll committed:
Send different size images + front end feedback

- Send increasingly better resolution images
- Waits for client to be done downloading before serving the latest
picture
- Remove unnecessary marker settings from constants for the new algo
- Will not serve pictures that are too similar
- Give center of detected marker instead of center of mass
- Use picture when it is ready instead of when the next picture is
taken
- Add documentaion strings for most functions and classes
- Rm some unused functions
- Changed loglevels
- Added "debug" mode to add verbose logging for the lid handler
- Make all pictures greyscale again
- `util.logtime()` a decorator that logs the time a function takes to
finish

amrsoll committed:
Add Analytics

- Analytics can now accept pretty much anything thanks to a new "json
serialisor" that takes care of difficult cases"

- Fixed a bug where the camera wouldn't stop, resulting in the
impossibility to restart octoprint

- Analytics when measuring the markers
  - Separated between the new and legacy algorithm
  - Tracks the average position of each of them
  - Logs the standart deviation between each measurement
  - Tracks how often a corner was found or missed
  - legacy and new algorithm can be compared 1:1

- Add new decorators and help functions in util/__init__.py

amrsoll committed:
Merge branch 'develop' into f_new_detection

amrsoll committed:
Send analytics when at least 1 pic was taken.

andywerner reviewed (changes requested):
we spoke about the todos

Marker detection ToDos:
- remember last know position of a marker (blocker for beta)
- slider: local maxima - check if it’s really a marker
- detected blobs: cycle is this a marker?
- reduce file size of the first image
- redo logic when markers flash on frontend
- Difference in file size between grey and colored.
- Timer to resend even with very slow connection
- use brightness for shutterspeed
- apply_best_shutter_speed() (maybe)

amrsoll committed:
Remember previous marker spot when covered

- Add comments from code review
- Give verbose errors to analytics with number of occurences
- Do not show uncorrected img if corners are not found
- JSON serialisation errors more verbose

TODO :

- Analytics : markers recovered from prev images should not count as
found
- Anti horizontal banding implementation for video feed

amrsoll committed:
Decrease quality on the first served picture

amrsoll commented:
- [x] remember last know position of a marker (blocker for beta)
- [ ] slider local maxima - check if it’s really a marker
- [ ] detected blobs: cycle is this a marker?
- [x] reduce file size of the first image
- [x] redo logic when markers flash on frontend
- [ ] use brightness for shutterspeed
- [ ] apply_best_shutter_speed() (maybe)
- [x] analytics documentation completed

Slow connection speeds (simulated regular 2G on Firefox):
First small picture served in 2 ~ 5 sec
Second bigger picture served in 14 sec

- [x] Difference in file size between grey and colored.

Low quality

- grayscale : 41 kB
- color : 45 kB

Better quality

- grayscale 418 kB
- color 492 kB


> - [x] Timer to resend even with very slow connection
> If front end hasn't shown sign that it is done downloading the last image, a new one will be sent.

amrsoll committed:
Merge remote-tracking branch 'origin/develop' into f_new_detection

amrsoll committed:
Bugfix : Redo front-end logic

- Circles now disappear when all the markers are found
- Compatible with new algorithm

amrsoll committed:
Serve pic when front ready + low light compensation.

- Let the raspberry wait more between 2 pictures
- Rm horizontal off set to the corner Regions of Interest (keep vertical
offset)

amrsoll committed:
revert debug state and set multicore

amrsoll committed:
Debug markers : Center marked with red crosshair
